ADR 026: Rosetta API Support (#6922)
cosmos/cosmos-sdk: docs/architecture/adr-026-rosetta-api.md

# ADR 26: Rosetta API Support

## Changelog

- August 3rd, 2020: Initial Draft.

## Status

- Approved, pending fine tuning the details of implementation.

## Context

> We think it'd be greatly valuable to application developers to have the Cosmos SDK

Reviewer: who is we? ADR's I don't think should specify something like `we`. Just needs reasoning why it will be a good idea

Reviewer 2: Typically, `we` is used in ADRs http://thinkrelevance.com/blog/2011/11/15/documenting-architecture-decisions.

Reviewer: ah okay, sorry I found it a bit odd, but ignore my comment

provide them with out-of-the-box Rosetta API support.
According to [the project's website](https://www.rosetta-api.org/), Rosetta API is an open
standard designed to simplify blockchain deployment and interaction. The latest specifications are
available at [this URL](https://www.rosetta-api.org/docs/Reference.html).

We want to achieve the following objectives:

- Support multiple versions of Cosmos SDK.
- Support Cosmos Hub.
- Implement querying of historical data sets.

## Decision

> We intend to develop a library that could be extended and used by application

Reviewer: I'm a bit lost. if its a seperate repo why is there an ADR in the sdk repo?

Author: The idea was first to be part of the same sdk repo, but @zmanian pointed out that there is one use case where we need to be able to query multiple versions of the hub at the same time, so living in its own repo is more "easy" to maintain than to have the same code in multiple versions of the SDK. 

But in the end is very related to the SDK so I opened the ADR here. But if you think is not correct to keep it here we can change it.

Reviewer: This ADR should be in the repo for where the code lives, IMO. Create repo -> add adr (design docs) to that repo -> code -> ship. What do others think?

Author: I think that is correct, meanwhile because the repo is not created yet (no permissions) we can use this as the draft discussion. Even I am not sure if finally everyone agrees on using another repo, this will be clarified in the cosmos-sdk call today I hope! Thanks @marbar3778

developers to integrate an in-process Rosetta API-compliant server with the
application main binaries. We also intend to provide a standalone gateway server

> program that supports a Cosmos SDK's minimum feature set. Such program could

Author: Thanks for pointing this out. 

I understand both ways are compatible as long as they live in the same Dockerfile (localhost or in-process), but if we had the Rosetta as external process in one Dockerfile and the node in another Dockerfile would not be possible.

run alongside the client applications main binaries.

> ### Implementation

Reviewer 2: I think a lot is missing from this section, particularly:

1. Mode specification.
2. How it's bootstrapped and started with a blockchain node (it should be via a sub-command of the app binary).
3. Types should be defined before their use (e.g. `Options` and `Router`)
4. What are `Options` and how are they used?
5. How are multiple networks supported from the same process/server?

```
type Server struct {
    options
    router
}

type Options struct {}

> func NewServer(opt Options,  rosettaAdapter RosettaAPIAdapter) {

Reviewer 3: ```suggestion
func NewServer(opt Options,  rosettaAdapter Adapter) {
```

    return Server{opt, newRouter(rosettaAdapter)}    
}
```

Server is the main type that the client application needs to instantiate. It embeds an `Options` type and exposes a `Start()` method:

```
func (s Server) Start() error {
    ...
    s.ListenAndServe(s.options.endpoint, s.router)
    ...
}
```

An internal `router` type exposes the Rosetta API endpoints:

```
type router struct {
    muxRouter mux.router

>     rosettaAdapter RosettaAPIAdapter

Reviewer 3: ```suggestion
    rosettaAdapter Adapter
```

}

> func newRouter(adapter RosettaAPIAdapter) {

Reviewer 3: ```suggestion
func newRouter(adapter Adapter) {
```

    router := &Router{
        mux: mux.NewRouter(),
        rosettaAdapter: adapter,
    }

    router.mux.HandleFunc("/blocks/", router.blocksApi)
    ...
    ...

    return router
}
```

Each API endpoint has an associated handler, e.g.:

```
func (r router) blocksApi(w http.ResponseWriter, r *http.Request) {
    // We build the RosettaGetBlockRequest
    req := ....

    res := r.rosettaAdapter.GetBlocks(req) 
    ...

    // We write the response. 
}
``` 

The router just translates a rest request into the data struct that the interface of an adapter needs to run.
The RosettaAdapter interface looks like this:

```

> type RosettaAdapterInterface interface {

Reviewer 3: ```suggestion
type Adapter interface {
```

    GetBlock(req RosettaGetBlockRequest) RosettaGetBlockResponse
    ...
}
```

> And we will provide different implementations of this `adapter` like for 0.38, for 0.39

Reviewer 3: ```suggestion
We will provide implementations of this `adapter` for each of the Cosmos SDK supported release series.
```

Author: In this case we can just create an Adapter of adapter that query different adapters and return the merged response. 

Query -> AdapterOfAdapters -> Adapter1 (cosmos-sdk 0.37) (Hub1)
                                                   -> Adapter2 (cosmos-sdk 0.38) (Hub2)
                                                   -> Adapter3 (cosmos-sdk 0.38) (Hub3)

The implementation maybe we need to offer for that is one for every SDK and a wrapper adapter of adapters to build complex use cases.

> and even a `CosmosHub` implementation that will call different versions of the hub.

Reviewer 3: ```suggestion
```

An example of an adapter implementation looks like this:

```
type CosmosLaunchpad struct {}

func (CosmosLaunchpad) GetBlock(req RosettaGetBlockRequest) {
    // Here we would just convert the RosettaGetBlockRequest 
    // into the data we need to get the block information
    // from a node that uses Launchpad.

    res := client.Get(...)

    // The adapter is responsable to build the response in the type
    // specified by the Interface.

    return RosettaGetBlockResponse{...}
}
...

```

The adapters that come out-of-the-box can be easily overriden by application-specific implementations.

This way we offer developers the opportunity to embed
a new server instance in their applications. Developers always retain the possiblity to using the simple binary

> that come with the Cosmos SDK.

Author: @alessio but this will be in its own repository I guess, so not in the Cosmos SDK.

Reviewer 3: Why? Can't we just build the binary from the SDK?

```
func (r router) blocksApi(w http.ResponseWriter, r *http.Request) {
    // We build the RosettaGetBlockRequest
    req := ....

    res := r.rosettaAdapter.GetBlocks(req) 
    ...

    // We write the response. 
}
``` 

The router just translates a rest request into the data struct that the interface of an adapter needs to run.
The RosettaAdapter interface looks like this:

```
type RosettaAdapterInterface interface {
    GetBlock(req RosettaGetBlockRequest) RosettaGetBlockResponse
    ...
}
```

And we will provide different implementations of this `adapter` like for 0.38, for 0.39
and even a `CosmosHub` implementation that will call different versions of the hub.

An example of an adapter implementation looks like this:

```
type CosmosLaunchpad struct {}

func (CosmosLaunchpad) GetBlock(req RosettaGetBlockRequest) {
    // Here we would just convert the RosettaGetBlockRequest 
    // into the data we need to get the block information
    // from a node that uses Launchpad.

    res := client.Get(...)

    // The adapter is responsable to build the response in the type
    // specified by the Interface.

    return RosettaGetBlockResponse{...}
}
...

```

> If our provided adapters don't work for a specific project because they changed their 
> API they can always create another adapter and inject it into the server.

Reviewer 3: ```suggestion
The adapters that come out-of-the-box can be easily overriden by application-specific implementations.
```

> This way we offer the possibility to offer developers the opportunity to instantiate

Reviewer 3: ```suggestion
This way we offer developers the opportunity to instantiate
```

> a new server in their applications or just using the binary that can be build from the

Reviewer 3: ```suggestion
a new server instance in their applications. Developers always retain the possiblity to using the simple binary
that come with the Cosmos SDK.
```

> repository.

Reviewer 3: ```suggestion
```

## Context

We think it'd be greatly valuable to application developers to have the Cosmos SDK
provide them with out-of-the-box Rosetta API support.
According to [the project's website](https://www.rosetta-api.org/), Rosetta API is an open
standard designed to simplify blockchain deployment and interaction. The latest specifications are
available at [this URL](https://www.rosetta-api.org/docs/Reference.html).

We want to achieve the following objectives:

- Support multiple versions of Cosmos SDK.
- Support Cosmos Hub.
- Implement querying of historical data sets.

## Decision

We intend to develop a library that could be extended and used by application
developers to integrate an in-process Rosetta API-compliant server with the
application main binaries. We also intend to provide a standalone gateway server
program that supports a Cosmos SDK's minimum feature set. Such program could
run alongside the client applications main binaries.

### Implementation

```
type Server struct {
    options
    router
}

type Options struct {}

func NewServer(opt Options,  rosettaAdapter RosettaAPIAdapter) {
    return Server{opt, newRouter(rosettaAdapter)}    
}
```

> Server is the main struct, it can be configured by an options struct.

Reviewer 3: ```suggestion
Server is the main type that the client application needs to instantiate. It embeds an `Options` type and exposes a `Start()` method:
```

> 
> This server exposes a Start method. 

Reviewer 3: ```suggestion
```

```
func (s Server) Start() error {
    ...
    s.ListenAndServe(s.options.endpoint, s.router)
    ...
}
```

> We have another struct which is the Router that will expose the rosetta endpoints.

Reviewer 3: ```suggestion
An internal `router` type exposes the Rosetta API endpoints:
```

```
type router struct {
    muxRouter mux.router
    rosettaAdapter RosettaAPIAdapter
}

func newRouter(adapter RosettaAPIAdapter) {
    router := &Router{
        mux: mux.NewRouter(),
        rosettaAdapter: adapter,
    }

    router.mux.HandleFunc("/blocks/", router.blocksApi)
    ...
    ...

    return router
}
```

> This router has a handler function by every Rosetta endpoint. The blocks handler looks like this:

Reviewer 3: ```suggestion
Each API endpoint has an associated handler., e.g.:
```

> # ADR 26: Rosetta API

Reviewer 3: ```suggestion
# ADR 26: Rosetta API Support
```

## Changelog

- August 3rd, 2020: Initial Draft.

## Status

- Approved, pending fine tuning the details of implementation.

## Context

> There is the need to provide an option to expose the Cosmos SDK applications to use
> the Rosetta API. This is an open Standard designed to simplify and integrate different
> blockchains.

Reviewer 3: ```suggestion
We think it'd be greatly valuable to application developers to have the Cosmos SDK
provide them with out-of-the-box Rosetta API support.
According to [the project's website](https://www.rosetta-api.org/), Rosetta API is an open
standard designed to simplify blockchain deployment and interaction. The latest specifications are
available at [this URL](https://www.rosetta-api.org/docs/Reference.html).
```

Our use case need to be able to provide:

- Support for the different versions of the Cosmos SDK.
- Support for different versions of the Cosmos Hub.
- Support to be able to provide an abstraction in order to provide querying data, let's say
we have a TxHash which we don't know if it is in cosmos hub 1, 2, 3. But queryng the Data

> API will return us the transaction without the need to query 3 versions.

Reviewer 3: We want to achieve the following objectives:

- Support multiple versions of Cosmos SDK.
- Support Cosmos Hub.
- Implement querying of historical data sets.

> Our use case need to be able to provide:
> 
> - Support for the different versions of the Cosmos SDK.
> - Support for different versions of the Cosmos Hub.
> - Support to be able to provide an abstraction in order to provide querying data, let's say
> we have a TxHash which we don't know if it is in cosmos hub 1, 2, 3. But queryng the Data
> API will return us the transaction without the need to query 3 versions.

Reviewer 3: ```suggestion
We want to achieve the following objectives:

- Support multiple versions of Cosmos SDK.
- Support Cosmos Hub.
- Implement querying of historical data sets.
```

## Decision

> In order to provide this use cases we have decided that the best approach will be to 
> use an external repository that will hold the libraries that can be integrated in
> different applications if they want to embed it into their own application and a
> main.go that will provide a binary that will run as a standalone.

Reviewer 3: ```suggestion
We intend to develop a library that could be extended and used by application
developers to integrate an in-process Rosetta API-compliant server with the
application main binaries. We also intend to provide a standalone gateway server
program that supports a Cosmos SDK's minimum feature set. Such program could
run alongside the client applications main binaries.
```

> This can be achieved providing a Server struct that is configured throught some options
> like:

Reviewer 3: ```suggestion
### Implementation
```
